[serve] Require prefix `RAY_SERVE_` for env vars + value verification

axreldable opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

- Require prefix `RAY_SERVE_` for all environment variable names in the `serve` module (excluding existent).
- Introduce deprecation warning and replacement for existing environment variable names without `RAY_SERVE_` prefix (deprecation in the next major `3.0.0` release):
  - the warning is logged only for explicitly defined variable
  - replacement contains `RAY_SERVE_` prefix (e.g. `RAY_SERVE_MAX_PER_REPLICA_RETRY_COUNT` for `MAX_PER_REPLICA_RETRY_COUNT`)

Commands:
```bash
ray start --head
serve config

# wrong name warning
export MAX_DEPLOYMENT_CONSTRUCTOR_RETRY_COUNT=1
export MAX_PER_REPLICA_RETRY_COUNT=1
export REQUEST_LATENCY_BUCKETS_MS=1
export MODEL_LOAD_LATENCY_BUCKETS_MS=1
export MAX_CACHED_HANDLES=1
export CONTROLLER_MAX_CONCURRENCY=1
export SERVE_REQUEST_PROCESSING_TIMEOUT_S=1
export RAY_SERVE_HANDLE_METRIC_PUSH_INTERVAL_S=1
```

Output:
```bash
bash$ serve config
~/ray/python/ray/serve/_private/constants.py:60: FutureWarning: Starting from version `3.0.0` environment variable `MAX_DEPLOYMENT_CONSTRUCTOR_RETRY_COUNT` will be deprecated. Please use `RAY_SERVE_MAX_DEPLOYMENT_CONSTRUCTOR_RETRY_COUNT` instead.
  get_env_int("MAX_DEPLOYMENT_CONSTRUCTOR_RETRY_COUNT", 20),
~/ray/python/ray/serve/_private/constants.py:67: FutureWarning: Starting from version `3.0.0` environment variable `MAX_PER_REPLICA_RETRY_COUNT` will be deprecated. Please use `RAY_SERVE_MAX_PER_REPLICA_RETRY_COUNT` instead.
  get_env_int("MAX_PER_REPLICA_RETRY_COUNT", 3),
~/ray/python/ray/serve/_private/constants.py:112: FutureWarning: Starting from version `3.0.0` environment variable `REQUEST_LATENCY_BUCKETS_MS` will be deprecated. Please use `RAY_SERVE_REQUEST_LATENCY_BUCKETS_MS` instead.
  get_env_str("REQUEST_LATENCY_BUCKETS_MS", ""),
~/ray/python/ray/serve/_private/constants.py:120: FutureWarning: Starting from version `3.0.0` environment variable `MODEL_LOAD_LATENCY_BUCKETS_MS` will be deprecated. Please use `RAY_SERVE_MODEL_LOAD_LATENCY_BUCKETS_MS` instead.
  get_env_str("MODEL_LOAD_LATENCY_BUCKETS_MS", ""),
~/ray/python/ray/serve/_private/constants.py:136: FutureWarning: Starting from version `3.0.0` environment variable `MAX_CACHED_HANDLES` will be deprecated. Please use `RAY_SERVE_MAX_CACHED_HANDLES` instead.
  "RAY_SERVE_MAX_CACHED_HANDLES", get_env_int_positive("MAX_CACHED_HANDLES", 100)
~/ray/python/ray/serve/_private/constants.py:143: FutureWarning: Starting from version `3.0.0` environment variable `CONTROLLER_MAX_CONCURRENCY` will be deprecated. Please use `RAY_SERVE_CONTROLLER_MAX_CONCURRENCY` instead.
  get_env_int_positive("CONTROLLER_MAX_CONCURRENCY", 15_000),
~/ray/python/ray/serve/_private/constants.py:255: FutureWarning: Starting from version `3.0.0` environment variable `SERVE_REQUEST_PROCESSING_TIMEOUT_S` will be deprecated. Please use `RAY_SERVE_REQUEST_PROCESSING_TIMEOUT_S` instead.
  get_env_float_non_negative("SERVE_REQUEST_PROCESSING_TIMEOUT_S", 0.0),
~/ray/python/ray/serve/_private/constants.py:306: FutureWarning: Starting from version `3.0.0` environment variable `RAY_SERVE_HANDLE_METRIC_PUSH_INTERVAL_S` will be deprecated. Please use `RAY_SERVE_HANDLE_AUTOSCALING_METRIC_PUSH_INTERVAL_S` instead.
  get_env_float("RAY_SERVE_HANDLE_METRIC_PUSH_INTERVAL_S", 10.0),
```

- Update `get_env_str` and `get_env_bool` to use internal `_get_env_value` + update type hints

--
- Replace `get_env_float_non_zero_with_warning` with `get_env_float_positive` for the `2.50.0` release. See PR #55464 for details. Now the following environment variables require positive value:
```
RAY_SERVE_PROXY_HEALTH_CHECK_TIMEOUT_S
RAY_SERVE_PROXY_HEALTH_CHECK_PERIOD_S
RAY_SERVE_PROXY_READY_CHECK_TIMEOUT_S
RAY_SERVE_PROXY_MIN_DRAINING_PERIOD_S
RAY_SERVE_KV_TIMEOUT_S
```

- Require non negative value for the following environment variables:
```
RAY_SERVE_HTTP_KEEP_ALIVE_TIMEOUT_S
RAY_SERVE_REQUEST_PROCESSING_TIMEOUT_S
RAY_SERVE_MULTIPLEXED_MODEL_ID_MATCHING_TIMEOUT_S
RAY_SERVE_QUEUE_LENGTH_CACHE_TIMEOUT_S
RAY_SERVE_MIN_HANDLE_METRICS_TIMEOUT_S
```

--
For this change, the following could be added to `2.50.0` **release notes**:

Ray Serve:
Enhancements:
Health check & env var safety:
- Introduced `RAY_SERVE_` required prefix for environment variables. 
- Warnings for invalid environment variable names, with migration path planned for Ray `3.0.0`.
- Added verification for timeout environment variables to be positive or non negative. 

<!-- Please give a short summary of the change and the problem this solves. -->

## Related issue number

<!-- For example: "Closes #1234" -->

Closes #55453

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a
           method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

axreldable committed:
[serve] Require prefix RAY_SERVE_ for env vars + deprecation warning

Signed-off-by: axreldable <aleksei.starikov.ax@gmail.com>

abrarsheikh commented:
Here’s the approach I’d suggest:

1. Create an allowlist of existing environment variables that currently violate the naming convention.
2. Introduce corrected versions of these variables with valid names, while keeping the old ones temporarily supported.
3. Warn users that the old variable names will be deprecated and removed in the next major release (Serve 3.0.0). Encourage them to update their configs proactively.
4. Enforce strict validation going forward: any new environment variables not in the allowlist must comply with the naming convention; otherwise they should fail the check.

We can’t immediately enforce the renaming of existing variables, since that would be too disruptive. These variables have been around for a long time, and forcing a change would risk breaking production setups when users upgrade.

axreldable commented:
/gemini review

axreldable commented:
/gemini review

axreldable commented:
Hi @abrarsheikh  , thank you for the review!

I’ve implemented your suggested approach. Could you please take a look and let me know if this aligns with what you had in mind?

axreldable committed:
[serve] Update naming + warn message

Signed-off-by: axreldable <aleksei.starikov.ax@gmail.com>

axreldable commented on python/ray/serve/_private/constants_utils.py:
```diff
@@ -47,13 +47,38 @@ def parse_latency_buckets(bucket_str: str, default_buckets: List[float]) -> List
 T = TypeVar("T")
 
+# todo: remove for the '3.0.0' release.
+_wrong_names_white_set = {
```
updated

axreldable commented on python/ray/serve/_private/constants_utils.py:
```diff
@@ -250,3 +287,33 @@ def get_env_float_non_zero_with_warning(
+            f"Got unexpected environment variable name `{name}`! "
+            f"Starting from version `{change_version}`, all environments "
+            f"variables related to `Ray Serve` will require prefix `{required_prefix}`. "
+            f"Change the name to start with the `{required_prefix}` prefix.",
```
Yes, good point, changed the warn message.

axreldable committed:
Merge branch 'master' into serve_env_warn

zcin commented on python/ray/serve/_private/constants_utils.py:
```diff
@@ -233,7 +270,7 @@ def get_env_float_non_zero_with_warning(
     RAY_SERVE_KV_TIMEOUT_S
 
-    todo: replace this function with 'get_env_float_positive' for the '2.50.0' release.
+    TODO: replace this function with 'get_env_float_positive' for the '2.50.0' release.
```
btw @axreldable 2.49 has been released, let's execute this todo for 2.50?

axreldable replied:
Hi @zcin! Thank you for raising it!  

I was wondering when do I need to do it. Now 2.49.1 is released. Should I prepare changes for the 2.50.0 release and merge to master now? I mean what if you decide to release 2.49.2? Will it intervene?

abrarsheikh replied:
Let's do it in this PR. Only way is subscribing to ray releases or watching the OSS slack channel. 

On a side note, if you are interested in engaging with more contributing please connect with me over slack and I can find the right scope for you :)

axreldable committed:
[serve] Remove get_env_float_non_zero_with_warning func from constants_utils.py

Signed-off-by: axreldable <aleksei.starikov.ax@gmail.com>

abrarsheikh reviewed (approved):
please leave a note on the PR description about what we should add to `release notes of 2.50`.

axreldable commented:
> please leave a note on the PR description about what we should add to `release notes of 2.50`.

For this change, the following could be added to `2.50.0` release notes:

Ray Serve:
Enhancements:
Health check & env var safety:
- Introduced `RAY_SERVE_` required prefix for environment variables. 
- Added verification for some of the timeout variable to be positive if specified. 
- Warnings for invalid environment variable names, with migration path planned for Ray `3.0.0`.

@abrarsheikh , do you think it makes sense? 

btw, while writing, I was wondering: should I require other timeout variables (with `_TIMEOUT_S` postfix) to be non-negative / positive if specified?

abrarsheikh commented:
> btw, while writing, I was wondering: should I require other timeout variables (with `_TIMEOUT_S` postfix) to be non-negative / positive if specified?

+1

And the release note looks good, we will use it during the next release.

axreldable committed:
[serve] Require non negative values for timeout env vars

Signed-off-by: axreldable <aleksei.starikov.ax@gmail.com>

axreldable commented:
> > btw, while writing, I was wondering: should I require other timeout variables (with `_TIMEOUT_S` postfix) to be non-negative / positive if specified?
> 
> +1
> 
> And the release note looks good, we will use it during the next release.

Added + updated description with the release notes.
